File: frontend/src/api/analyses.ts
import { apiClient } from "../lib/apiClient";
import { supabase } from "../lib/supabase";

export interface CreateAnalysisInput {
  resume: File;
  jdText: string;
  jobTitle?: string;
  company?: string;
}

export interface CreateAnalysisResponse {
  analysis_id: string;
  resume_id: string;
  status: string;
}

export interface AnalysisResult {
  score?: number;
  requirements?: string;
  must_skills?: string[];
  nice_skills?: string[];
  matched_skills?: string[];
  missing_skills?: string[];
  suggestions?: string;
  rewritten_resume?: string;
  steps?: Array<{ step: string; status: string }>;
}

export interface AnalysisEvent {
  type: "progress" | "step" | "completed" | "failed";
  analysis_id?: string;
  step?: string;
  status?: string;
  progress?: number;
  score?: number;
  message?: string;
  content?: unknown;
  result?: AnalysisResult;
}

export interface AnalysisListItem {
  id: string;
  resume_id: string;
  status: string;
  score: number | null;
  progress: number;
  current_step: string | null;
  job_title: string | null;
  company: string | null;
  created_at: string;
  updated_at: string;
}

export async function createAnalysis(
  input: CreateAnalysisInput,
  requestId: string,
): Promise<CreateAnalysisResponse> {
  const formData = new FormData();
  formData.append("resume", input.resume);
  formData.append("jd_text", input.jdText);
  if (input.jobTitle) formData.append("job_title", input.jobTitle);
  if (input.company) formData.append("company", input.company);

  const response = await apiClient.post("/api/v1/analyses", formData, {
    headers: { "X-Request-ID": requestId },
  });
  return response.data.data;
}

export async function listAnalyses(): Promise<AnalysisListItem[]> {
  const response = await apiClient.get("/api/v1/analyses");
  return response.data.data.items;
}

export function createAnalysisEventsUrl(analysisId: string) {
  const baseURL = apiClient.defaults.baseURL ?? "";
  return `${baseURL}/api/v1/analyses/${analysisId}/events`;
}

export async function streamAnalysisEvents(
  analysisId: string,
  onEvent: (event: AnalysisEvent) => void,
  signal?: AbortSignal,
) {
  const { data } = await supabase.auth.getSession();
  const token = data.session?.access_token;
  const headers: HeadersInit = {};
  if (token) headers.Authorization = `Bearer ${token}`;

  const response = await fetch(createAnalysisEventsUrl(analysisId), {
    headers,
    signal,
  });
  if (!response.ok) {
    throw new Error("Failed to connect to analysis event stream");
  }
  if (!response.body) {
    throw new Error("Analysis event stream is unavailable");
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  const processBlock = (block: string) => {
    const dataLines = block
      .split("\n")
      .filter((line) => line.startsWith("data: "))
      .map((line) => line.slice(6));
    if (!dataLines.length) return;
    onEvent(JSON.parse(dataLines.join("\n")) as AnalysisEvent);
  };

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const blocks = buffer.split("\n\n");
    buffer = blocks.pop() ?? "";
    for (const block of blocks) {
      processBlock(block);
    }
  }

  if (buffer.trim()) {
    processBlock(buffer);
  }
}

File: backend/routers/analyses.py
import json

from fastapi import APIRouter, Depends, Form, Request, UploadFile
from fastapi.responses import StreamingResponse

from controllers.analysis_controller import (
    create_analysis_from_upload,
    delete_analysis,
    get_analysis,
    list_analyses,
)
from core.responses import api_error, api_success
from core.security import get_current_user
from services.analysis_stream_service import stream_hub

router = APIRouter(prefix="/analyses")


@router.post("")
async def create_analysis(
    request: Request,
    resume: UploadFile,
    jd_text: str = Form(...),
    job_title: str | None = Form(None),
    company: str | None = Form(None),
    current_user: dict = Depends(get_current_user),
):
    try:
        return await create_analysis_from_upload(
            current_user=current_user,
            request_id=request.state.request_id,
            resume=resume,
            jd_text=jd_text,
            job_title=job_title,
            company=company,
        )
    except ValueError as exc:
        return api_error(str(exc))


@router.get("")
async def get_analyses(current_user: dict = Depends(get_current_user)):
    return api_success({"items": list_analyses(current_user=current_user)})


@router.get("/{analysis_id}")
async def get_analysis_detail(analysis_id: str, current_user: dict = Depends(get_current_user)):
    analysis = get_analysis(current_user=current_user, analysis_id=analysis_id)
    if not analysis:
        return api_error("Analysis not found")
    return api_success(analysis)


@router.delete("/{analysis_id}")
async def delete_analysis_detail(analysis_id: str, current_user: dict = Depends(get_current_user)):
    delete_analysis(current_user=current_user, analysis_id=analysis_id)
    return api_success({"deleted": True})


@router.get("/{analysis_id}/events")
async def analysis_events(
    analysis_id: str,
    current_user: dict = Depends(get_current_user),
):
    analysis = get_analysis(current_user=current_user, analysis_id=analysis_id)
    snapshot_event = _analysis_snapshot_event(analysis) if analysis else {
        "type": "failed",
        "analysis_id": analysis_id,
        "message": "Analysis not found",
    }

    async def generate():
        yield _sse_event(snapshot_event)
        if snapshot_event.get("type") in {"completed", "failed"}:
            return

        async for event in stream_hub.subscribe(analysis_id):
            yield _sse_event(event)

    return StreamingResponse(generate(), media_type="text/event-stream")


def _sse_event(event: dict) -> str:
    return f"event: {event.get('type', 'progress')}\n" f"data: {json.dumps(event)}\n\n"


def _analysis_snapshot_event(analysis: dict) -> dict:
    analysis_id = str(analysis["id"])
    status = analysis["status"]

    if status == "completed":
        return {
            "type": "completed",
            "analysis_id": analysis_id,
            "status": "completed",
            "progress": 100,
            "score": analysis.get("score"),
            "message": "Analysis completed",
            "result": analysis.get("result") or {},
        }

    if status == "failed":
        return {
            "type": "failed",
            "analysis_id": analysis_id,
            "status": "failed",
            "progress": 100,
            "message": analysis.get("error") or "Analysis failed",
        }

    return {
        "type": "progress",
        "analysis_id": analysis_id,
        "step": analysis.get("current_step") or status,
        "status": status,
        "progress": analysis.get("progress") or 0,
        "message": f"Analysis is {status}",
    }
